File: data_analysis/Nterm_13mer_sequence_features.R
# import packages
library(tidyverse)

# for executing python script in Rstudio
library(reticulate)

# use specific virtual env created by anaconda
use_condaenv(
  condaenv = '/opt/anaconda3/envs/Nterm_probe',
  required = TRUE
)

# execute the python script for calculating the sequence features
source_python("data_analysis/Nterm_13mer_sequence_features.py")

# import the result from localCIDER (data_analysis/Nterm_sequence_features.py)
HEK_Nterm_13mer_sequence_features <- read_csv(
  'data_source/Nterm_13mer_sequence_features/HEK_Nterm_13mer_Kd_half_life_sequence_features.csv'
)

## Wilcoxon rank-sum test
library(rstatix)

# hydropathy
HEK_Nterm_13mer_sequence_features |> 
  wilcox_test(hydropathy ~ category)

# FCR
HEK_Nterm_13mer_sequence_features |> 
  wilcox_test(FCR ~ category)

# NCPR
HEK_Nterm_13mer_sequence_features |> 
  wilcox_test(NCPR ~ category)

# isoelectric point
HEK_Nterm_13mer_sequence_features |> 
  wilcox_test(isoelectric_point ~ category)

# kappa
HEK_Nterm_13mer_sequence_features |> 
  wilcox_test(kappa ~ category)

File: data_analysis/Nterm_13mer_sequence_features.py
# import packages
import pandas as pd
from tqdm import tqdm

# import localCIDER package (https://pappulab.github.io/localCIDER/, Version 0.1.20)
# using for calculate various properties of a peptide sequence
from localcider.sequenceParameters import SequenceParameters

# import Nterm 13mer sequence
HEK_Nterm_Kd_half_life_sequence = pd.read_csv(
  'data_source/Nterm_sequence/HEK_Nterm_Kd_half_life_sequence.csv',
  header = 0,
  index_col = None
)

# get sequence parameters for Nterm_13mer
HEK_Nterm_Kd_half_life_sequence.loc[:, "Nterm_13mers_sequence_parameter"]= HEK_Nterm_Kd_half_life_sequence["Nterm_13mer"].astype(pd.StringDtype()).apply(SequenceParameters)

## calculate hydropathy, FCR, NCPR, isoelectric point and kappa values
methods = {
    "hydropathy": "get_mean_hydropathy",
    "FCR": "get_FCR",
    "NCPR": "get_NCPR",
    "isoelectric_point": "get_isoelectric_point",
    "kappa": "get_kappa"
}

for col, method in tqdm(methods.items()):
    HEK_Nterm_Kd_half_life_sequence.loc[:, col] = HEK_Nterm_Kd_half_life_sequence["Nterm_13mers_sequence_parameter"].apply(
        lambda obj: getattr(obj, method)() if obj is not None else None
    )

# export result
HEK_Nterm_Kd_half_life_sequence.to_csv(
  "data_source/Nterm_13mer_sequence_features/HEK_Nterm_13mer_Kd_half_life_sequence_features.csv", 
  index = False
)
